pkg/gatewayconfigurationserver/grpc.go

```diff
@@ -0,0 +1,127 @@
+	"go.thethings.network/lorawan-stack/v3/pkg/frequencyplans"
+	"go.thethings.network/lorawan-stack/v3/pkg/pfconfig/cpf"
+	"go.thethings.network/lorawan-stack/v3/pkg/pfconfig/semtechudp"
+	"go.thethings.network/lorawan-stack/v3/pkg/ttnpb"
+)
+
+var errInvalidConfigurationFormat = errors.DefineInvalidArgument("invalid_configuration_format", "configuration format `{format}` is not supported")
+var errInvalidConfigurationType = errors.DefineInvalidArgument("invalid_configuration_type", "configuration type `{type}` is not supported")
+var errInvalidConfigurationFilename = errors.DefineInvalidArgument("invalid_configuration_filename", "configuration filename `{filename}` is not supported")
+
+func (s *Server) getGateway(ctx context.Context, gtwID *ttnpb.GatewayIdentifiers) (*ttnpb.Gateway, error) {
+	cc, err := s.GetPeerConn(ctx, ttnpb.ClusterRole_ENTITY_REGISTRY, nil)
+	if err != nil {
+		return nil, err
+	}
+	client := ttnpb.NewGatewayRegistryClient(cc)
+	gtw, err := client.Get(ctx, &ttnpb.GetGatewayRequest{
+		GatewayIds: gtwID,
+		FieldMask: &types.FieldMask{
+			Paths: []string{
+				"antennas",
+				"frequency_plan_id",
+				"gateway_server_address",
+			},
+		},
+	}, s.WithClusterAuth())
+	if err != nil {
+		return nil, err
+	}
+	return gtw, nil
+}
+
+// GetGatewayConfiguration validates the request fields and returns the appropriate gateway configuration
+func (s *Server) GetGatewayConfiguration(ctx context.Context, req *ttnpb.GetGatewayConfigurationRequest) (*ttnpb.GetGatewayConfigurationResponse, error) {
+	if s.config.RequireAuth {
+		err := rights.RequireGateway(ctx, *req.GatewayIds, ttnpb.Right_RIGHT_GATEWAY_INFO)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	gtw, err := s.getGateway(ctx, req.GetGatewayIds())
+	if err != nil {
+		return nil, err
+	}
+	fps, err := s.FrequencyPlansStore(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	var configContent []byte
+	switch req.Format {
+	case "semtechudp":
+		configContent, err = handleSemtechUDP(gtw, fps, req)
+
+	case "kerlink-cpf":
+		configContent, err = handleKerlinkCPF(gtw, fps, req)
+
+	default:
+		return nil, errInvalidConfigurationFormat.WithAttributes("format", req.Format)
+	}
+
+	return &ttnpb.GetGatewayConfigurationResponse{
+		Contents: configContent,
+	}, err
+}
+
+func handleSemtechUDP(gtw *ttnpb.Gateway, fps *frequencyplans.Store, req *ttnpb.GetGatewayConfigurationRequest) ([]byte, error) {
+	if req.Filename != "global_conf.json" {
+		return nil, errInvalidConfigurationFilename.WithAttributes("filename", req.Filename)
+	}
+	config, err := semtechudp.Build(gtw, fps)
+	if err != nil {
+		return nil, err
+	}
+	return json.Marshal(config)
+}
+
+func handleKerlinkCPF(gtw *ttnpb.Gateway, fps *frequencyplans.Store, req *ttnpb.GetGatewayConfigurationRequest) ([]byte, error) {
+	switch req.Type {
+	case "lorad":
+		if req.Filename != "lorad.json" {
+			return nil, errInvalidConfigurationFilename.WithAttributes("filename", req.Filename)
```

Comment:
I think it would be good to also log the type here.
```
configuration filename `{filename}` for type `{type}` is not supported
```

Also in some of the other log statements.